test/common/http/http2/metadata_encoder_decoder_test.cc

```diff
@@ -322,31 +205,116 @@ TEST_F(MetadataEncoderDecoderTest, VerifyEncoderDecoderMultipleMetadataReachSize
     memset(output_buffer_.buf, 0, output_buffer_.length);
     output_buffer_.length = 0;
 
-    MetadataMap metadata_map_2 = {
+    MetadataMap metadata_map = {
         {"header_key1", std::string(10000, 'a')},
         {"header_key2", std::string(10000, 'b')},
     };
+    MetadataMapVec metadata_map_vec;
+    metadata_map_vec.push_back(&metadata_map);
 
     // Encode and decode the second MetadataMap.
-    MetadataCallback cb2 = std::bind(&MetadataEncoderDecoderTest::verifyMetadata, this,
-                                     metadata_map_2, std::placeholders::_1);
+    MetadataCallback cb2 = std::bind(&MetadataEncoderDecoderTest::verifyMetadataMapVec, this,
+                                     metadata_map_vec, std::placeholders::_1);
     decoder_->callback_ = cb2;
-    encoder_.createPayload(metadata_map_2);
-    nghttp2_submit_extension(session_, METADATA_FRAME_TYPE, END_METADATA_FLAG, STREAM_ID, nullptr);
-    nghttp2_session_send(session_);
+    submitMetadata(metadata_map_vec);
+
     result = nghttp2_session_mem_recv(session_, output_buffer_.buf, output_buffer_.length);
     if (result < 0) {
       break;
     }
   }
-
   // Verifies max matadata limit reached.
   EXPECT_LT(result, 0);
   EXPECT_LE(decoder_->max_payload_size_bound_, decoder_->total_payload_size_);
 
   cleanUp();
 }
 
+// Tests encoding/decoding small metadata map vectors.
+TEST_F(MetadataEncoderDecoderTest, EncodeMetadataMapVecSmall) {
+  MetadataMap metadata_map = {
+      {"header_key1", std::string(5, 'a')},
+      {"header_key2", std::string(5, 'b')},
+  };
+  MetadataMap metadata_map_2 = {
+      {"header_key3", std::string(5, 'a')},
+      {"header_key4", std::string(5, 'b')},
+  };
+  MetadataMap metadata_map_3 = {
+      {"header_key1", std::string(1, 'a')},
+      {"header_key2", std::string(1, 'b')},
+  };
+
+  MetadataMapVec metadata_map_vec;
+  metadata_map_vec.push_back(&metadata_map);
+  metadata_map_vec.push_back(&metadata_map_2);
+  metadata_map_vec.push_back(&metadata_map_3);
+
+  // Verifies the encoding/decoding result in decoder's callback functions.
+  MetadataCallback cb = std::bind(&MetadataEncoderDecoderTest::verifyMetadataMapVec, this,
+                                  metadata_map_vec, std::placeholders::_1);
+  initialize(cb);
+  submitMetadata(metadata_map_vec);
+
+  // Verifies flag and payload are encoded correctly.
+  const uint64_t consume_size = random_generator_.random() % output_buffer_.length;
+  nghttp2_session_mem_recv(session_, output_buffer_.buf, consume_size);
+  nghttp2_session_mem_recv(session_, output_buffer_.buf + consume_size,
+                           output_buffer_.length - consume_size);
+
+  cleanUp();
+}
+
+// Tests encoding/decoding large metadata map vectors.
+TEST_F(MetadataEncoderDecoderTest, EncodeMetadataMapVecLarge) {
```

Comment:
Along with this do you think it might be useful to have a test which fuzzed frame boundaries 3-4 vectors which had random length values?  I think these tests cover most of what we want but if your logic had an off by one error in frame calculation I suspect I probably wouldn't catch it via visual inspection :-/

Reply:
Good point! Fixed.